src/Microsoft.Identity.Web/WebAppServiceCollectionExtensions.cs

```diff
@@ -40,20 +47,60 @@ public static class WebAppServiceCollectionExtensions
-               return builder.AddSignIn(
-                    options => configuration.Bind(configSectionName, options),
-                    options => configuration.Bind(configSectionName, options),
-                    openIdConnectScheme,
-                    cookieScheme,
-                    subscribeToOpenIdConnectMiddlewareDiagnosticsEvents);
+            builder.AddSignIn(
+                options => configuration.Bind(configSectionName, options),
+                options => configuration.Bind(configSectionName, options),
+                openIdConnectScheme,
+                cookieScheme,
+                subscribeToOpenIdConnectMiddlewareDiagnosticsEvents);
+            return services;
+        }
+
+        /// <summary>
+        /// Add authentication with Microsoft identity platform.
+        /// This method expects the configuration file will have a section, named "AzureAd" as default, with the necessary settings to initialize authentication options.
+        /// </summary>
+        /// <param name="services">Service collection to which to add authentication</param>
+        /// <param name="configureMicrosoftIdentityOptions">the action to set the <see cref="MicrosoftIdentityOptions"/></param>
+        /// <param name="configureMicrosoftIdentityOptions">the action to set the <see cref="MicrosoftIdentityOptions"/></param>
+        /// <param name="openIdConnectScheme">Optional name for the open id connect authentication scheme 
+        /// (by default OpenIdConnectDefaults.AuthenticationScheme). This can be specified when you want to support
+        /// several OpenIdConnect identity providers.</param>
+        /// <param name="cookieScheme">Optional name for the acookie uthentication scheme 
+        /// (by default OpenIdConnectDefaults.AuthenticationScheme) </param>
+        /// <param name="subscribeToOpenIdConnectMiddlewareDiagnosticsEvents">
+        /// Set to true if you want to debug, or just understand the OpenIdConnect events.
+        /// </param>
+        /// <returns>Yhe service collection for chaining</returns>
+        public static IServiceCollection AddSignIn(
+            this IServiceCollection services,
+            Action<OpenIdConnectOptions> configureOpenIdConnectOptions,
+            Action<MicrosoftIdentityOptions> configureMicrosoftIdentityOptions,
+            string openIdConnectScheme = OpenIdConnectDefaults.AuthenticationScheme,
+            string cookieScheme = CookieAuthenticationDefaults.AuthenticationScheme,
+            bool subscribeToOpenIdConnectMiddlewareDiagnosticsEvents = false)
+        {
+            AuthenticationBuilder builder = services.AddAuthentication(openIdConnectScheme);
+            builder.AddSignIn(
+                 configureOpenIdConnectOptions,
+                 configureMicrosoftIdentityOptions,
+                 openIdConnectScheme,
+                 cookieScheme,
+                 subscribeToOpenIdConnectMiddlewareDiagnosticsEvents);
+            return services;
```

Comment:
I think we need to return `builder` object here, which actually has the `Services` property in itself. From what I've seen in the ASP.NET Core `IServiceCollection` extensions classes, if the extension method only adds some services to the collection, then it's fine to return `services`, but if there's more setup using a builder, then builder is returned. Also `AuthenticaitonBuilder` has some specific auth related methods that can be chained. For example:
```CSharp
services.AddAuthentication(X).AddSignIn(Y).AddCookie(Z).Services.AddLogging();
```

Reply:
like the idea. let's dicuss

Reply:
I get your point, @pmaytak. Thanks for this feedback.

Today, by returning services, we enable (simple) code like:

```csharp
services.AddSignIn(Configuration)
             .AddProtectedWebApiCallsWebAPi(Configuration)
             .AddTokenCacheXXX();
```

If we returned a builder, the code would have to be

```csharp
services.AddSignIn(Configuration);
services.AddProtectedWebApiCallsWebAPi(Configuration)
             .AddTokenCacheXXX();
```

which is ok, but I like a bit less, and given we are already taking care of cookies ...

Finally, it's always possible for customers who want to get a builder to use the following override:

```csharp
services.AddAuthentication(OpenIdConnectAuthentication.Default)
             .AddSignIn(Configuration)
             .AddCookie(Y);

services.AddProtectedWebApiCallsWebAPi(Configuration)
             .AddTokenCacheXXX();
```

Reply:
If we returned a builder, your second code example could be this:
```CSharp
services.AddSignIn(Configuration);
        .Services
        .AddProtectedWebApiCallsWebAPi(Configuration)
        .AddTokenCacheXXX();
```
Similar to code in [ServiceCollectionTests.cs](https://github.com/dotnet/aspnetcore/blob/43c2c9ea97dfacce87f6bd578c1863ee0bf677c4/src/DataProtection/DataProtection/test/ServiceCollectionTests.cs#L15)

I agree with your point of making the API simple, and both of the ways are pretty close in that regard. The builder does provide more authentication setup functionality while still providing access to the `Services` property for chaining. I guess my perspective is a bit more conceptual - do we return a more specific `AuthenticationBuilder` or a more general `IServicesCollection`.

Reply:
Proposing to do a user study on that
(ask users what they prefer)?